### DISCUSSION TITLE
combine rectangular and cylindrical grids/meshes

### DISCUSSION THREAD

**Discussion by woecj** [thliebig/openEMS-Project#226] [Q&A Usage]
Hi,
I want to analyse a structure having a rectangular waveguide and cylindrical elements.
How to build a combined mesh - rectangular for the waveguide and cylindrical for a round plate?
Thanks,
Johann

**Comment by LubomirJagos42** [comment]
hi, in openEMS you can't combine rectangular and cylindrical mesh, you can use just one for whole simulation
therefore in your case you should use rectangular mesh since waveguide is rectangular and for cylindrical elements you have to setup quite fine mesh to got right objects representation in simulation.
I never used but there are multiple option for meshing like subgrids and so, so you should look on that.

sorry there is no way to combine this

**Reply by thliebig** [reply]
The subgrids are only available for cylindrical meshes and only to reduce the mesh density around the cylinder center.

**Reply by woecj** [reply]
Thank you for your answers.
A am a new user of openEMS, so I fight against beginners missing knowledge on use.
The device I want to simulate is a rectangular waveguide with a thin disc on a small distance to one end of the waveguide perpendicular to the axis.
Up to now I could not find a solution to mesh the disc at necessary tightness without also mesh the waveguide with the same grid spacing.
How can I achieve that?

**Reply by VolkerMuehlhaus** [reply]
Welcome to FDTD ;-)

The mesh lines that you need for these small geometries extend throughout the entire analysis volume, there is no other option.

That said, you don't need to fill the entire **volume** with the same small **cube**. Just place mesh lines to resolve your geometry and increase the mesh density up to lambda/20 in "empty" areas where no geometry details need to be resolved.

![extra_meshlines](https://github.com/user-attachments/assets/034d553f-13d3-4391-a079-66293a1cc637)

**Comment by woecj** [comment]
thank you for your hints!
There are functions to support meshing on entities, for example DetectEdges(....), but I could not find any documentation on their use, just some use of them in the tutorials (Simple_Patch_Antenna.m).
Maybe they could help me.

**Reply by VolkerMuehlhaus** [reply]
I am afraid that it is beyond the scope of this forum to teach efficient meshing. Meshing is the most diffcult part in openEMS, and has a strong effect on results and simulation time.

My knowledge about "good" FDTD meshing strategy is from using a commercial FDTD tool with automatic meshing, and looking at the resulting mesh. openEMS has no automatic meshing, so my openEMS models require hand-made (hand coded) mesh to achieve a similar end result. Basically it is "fine enough where it matters", refined near the conductor edges.

You might not want to hear this, but be prepared to spend some weeks on studying examples, to get an idea what a good mesh looks like. Once you get the idea, you can start coding an optimized mesh for your own geometries. 

I wish I had a better answer, but free EM tools require a lot of time to get started. I'm currently investigating a free FEM code and the first few weeks were really frustrating, although I have 25+ years of EM experience with commercial EM solvers ;-)

**Reply by VolkerMuehlhaus** [reply]
Reagrding documentation: also have a look at the *.m and *.py source code here in github, depending on what language for openEMS you are using. The code has a lot of useful comments inside, that helped me a lot!

**Reply by woecj** [reply]
Thank you for the quick answers and valuable informations.
Seems, I have to investigate quite more time to get reliable results from this very powerful tool.
